File: python/playwright-ad-block/scrape.py
#!/usr/bin/env python3
from os import environ
from time import time
from playwright.sync_api import sync_playwright

AUTH = environ.get('AUTH', default='USER:PASS')
TARGET_URL = environ.get('TARGET_URL', default='https://www.w3schools.com/js/js_dates.asp')
AD_BLOCK = environ.get('AD_BLOCK', default='1')

def scrape(url=TARGET_URL):
    if AUTH == 'USER:PASS':
        raise Exception('Provide Scraping Browsers credentials in AUTH '
                        'environment variable or update the script.')
    print('Connecting to Browser...')
    browser_ws_endpoint = f'wss://{AUTH}@brd.superproxy.io:9222'
    with sync_playwright() as playwright:
        browser = playwright.chromium.connect_over_cdp(browser_ws_endpoint)
        try:
            begin = time()
            print('Connected!')
            page = browser.new_page()
            client = page.context.new_cdp_session(page)
            result = client.send('Browser.getSessionId')
            session_id = result['sessionId']
            print('Current session ID:', session_id)
            if AD_BLOCK == '1':
                print('Enabling AdBlock...')
                client = page.context.new_cdp_session(page)
                client.send('Unblocker.enableAdBlock')
                print('Enabled!')
            print(f'Navigating to {url}...')
            page.goto(url, timeout=2 * 60 * 1000)

            data = page.content()
            elapsed = time() - begin
            print(f'Done! Time: {elapsed}')
            return data

        finally:
            browser.close()


if __name__ == '__main__':
    scrape()

File: nodejs/playwright-ad-block/scrape.js
#!/usr/bin/env node
const playwright = require('playwright');
const {
    AUTH = 'USER:PASS',
    TARGET_URL = 'https://www.w3schools.com/js/js_dates.asp',
    AD_BLOCK = '1',
} = process.env;

async function scrape(url = TARGET_URL) {
    if (AUTH == 'USER:PASS') {
        throw new Error(`Provide Scraping Browsers credentials in AUTH`
            + ` environment variable or update the script.`);
    }
    console.log(`Connecting to Browser...`);
    const browserWSEndpoint = `wss://${AUTH}@brd.superproxy.io:9222`;
    const browser = await playwright.chromium.connectOverCDP(browserWSEndpoint);
    try {
        const begin = Date.now();
        console.log(`Connected!`);
        const page = await browser.newPage();
        const client = await page.context().newCDPSession(page);
        const result = await client.send('Browser.getSessionId');
        const sessionId = result.sessionId;
        console.log('Current session ID:', sessionId);
        if (AD_BLOCK == '1') {
            console.log(`Enabling AdBlock...`);
            await client.send('Unblocker.enableAdBlock');
            console.log(`Enabled!`);
        }
        console.log(`Navigating to ${url}...`);
        await page.goto(url, { timeout: 2 * 60 * 1000 });

        const data = await page.content();
        const time = Date.now() - begin;
        console.log(`Done! Time: ${time}`);
        return data;

    } finally {
        await browser.close();
    }
}

function getErrorDetails(error) {
    if (error.target?._req?.res) {
        const {
            statusCode,
            statusMessage,
        } = error.target._req.res;
        return `Unexpected Server Status ${statusCode}: ${statusMessage}`;
    }
}

if (require.main == module) {
    scrape().catch(error => {
        console.error(getErrorDetails(error)
            || error.stack
            || error.message
            || error);
        process.exit(1);
    });
}
